src/zigTestRunnerProvider.ts

```diff
@@ -0,0 +1,204 @@
+import vscode from "vscode";
+
+import childProcess from "child_process";
+import path from "path";
+import util from "util";
+
+import { DebouncedFunc, throttle } from "lodash-es";
+
+import { getWorkspaceFolder, getZigPath, isWorkspaceFile } from "./zigUtil";
+
+const execFile = util.promisify(childProcess.execFile);
+
+export default class ZigTestRunnerProvider {
+    private testController: vscode.TestController;
+    private updateTestItems: DebouncedFunc<(document: vscode.TextDocument) => void>;
+
+    constructor() {
+        this.updateTestItems = throttle(
+            (document: vscode.TextDocument) => {
+                this._updateTestItems(document);
+            },
+            500,
+            { trailing: true },
+        );
+
+        this.testController = vscode.tests.createTestController("zigTestController", "Zig Tests");
+        this.testController.createRunProfile("Run", vscode.TestRunProfileKind.Run, this.runTests.bind(this), true);
+        this.testController.createRunProfile(
+            "Debug",
+            vscode.TestRunProfileKind.Debug,
+            this.debugTests.bind(this),
+            false,
+        );
+        void this.findAndRegisterTests();
+    }
+
+    public activate(subscriptions: vscode.Disposable[]) {
+        subscriptions.push(
+            vscode.workspace.onDidOpenTextDocument((document) => {
+                this.updateTestItems(document);
+            }),
+            vscode.workspace.onDidCloseTextDocument((document) => {
+                !isWorkspaceFile(document.uri.fsPath) && this.deleteTestForAFile(document.uri);
+            }),
+            vscode.workspace.onDidChangeTextDocument((change) => {
+                this.updateTestItems(change.document);
+            }),
+            vscode.workspace.onDidDeleteFiles((event) => {
+                event.files.forEach((file) => {
+                    this.deleteTestForAFile(file);
+                });
+            }),
+            vscode.workspace.onDidRenameFiles((event) => {
+                event.files.forEach((file) => {
+                    this.deleteTestForAFile(file.oldUri);
+                });
+            }),
+        );
+    }
+
+    private deleteTestForAFile(uri: vscode.Uri) {
+        this.testController.items.forEach((item) => {
+            if (!item.uri) return;
+            if (item.uri.fsPath === uri.fsPath) {
+                this.testController.items.delete(item.id);
+            }
+        });
+    }
+
+    private async findAndRegisterTests() {
+        const files = await vscode.workspace.findFiles("**/*.zig");
+        for (const file of files) {
+            try {
+                const doc = await vscode.workspace.openTextDocument(file);
+                this._updateTestItems(doc);
+            } catch {}
+        }
+    }
+
+    private _updateTestItems(textDocument: vscode.TextDocument) {
+        if (textDocument.languageId !== "zig") return;
+
+        const regex = /\btest\s+"([^"]+)"\s*\{/g;
```

Comment:
This regex doesn't work for [doctests](https://ziglang.org/documentation/master/#Doctests). Something like `/\btest\s*(?:@?"([^"]+)"|([a-zA-Z_][\w]*))\s*\{/g` should but it then has issues with multiple doctests in a row.

```zig
test foo {} // no run button
test "foo" {} // run button
test bar {} // no run button
test baz {} // no run button
test quux {} // run button
```

Reply:
Hm.. I didn't even notice we had doctests :). But I updated the regex and a code a bit to account for these type of test. 

There is one simple scenario I don't know if we can solve though.

e.g we have two tests 
```zig
test foo{
  // passing test
}

test "foo"{
  // failing test
}
```
Because these both have technically same name, and zig `--test-filter` can't distinguish between these two - the result for the test runs are shared unfortunately.